frontend/z3_types.py

```diff
@@ -0,0 +1,355 @@
+    quantifiers_consts = [Const("tuples_q_{}".format(x), type_sort) for x in range(len(tuples) - 1)]
+    axioms = []
+    x = Const("x", type_sort)
+    for i in range(len(tuples)):
+        # Tuple tuples[i] will have length i
+        if i == 0:
+            # Zero length tuple: An expression not a call.
+            axioms.append(extends(tuples[i], type_sort.tuple))
+            axioms.append(ForAll(x, Implies(subtype(x, tuples[i]), x == tuples[i])))
+        else:
+            # Tuple tuples[i] uses exactly i constants
+            consts = quantifiers_consts[:i]
+            inst = tuples[i](consts)
+            axioms.append(
+                ForAll(consts, extends(inst, type_sort.tuple), patterns=[inst])
+            )
+            axioms.append(ForAll([x] + consts, Implies(subtype(x, inst), x == inst)))
+
+    return axioms
+
+
+def functions_subtype_axioms(funcs, type_sort):
+    """Add the axioms for the functions subtyping"""
+
+    # Initialize constants to be used in the ForAll quantifier
+    # Each function needs a number of quantifiers equal to its args length + 1 (for the return type).
+    # With n functions, from zero-length to (n - 1) length arguments, we need at most n quantifiers.
+    quantifiers_consts = [Const("funcs_q_{}".format(x), type_sort) for x in range(len(funcs))]
+    axioms = []
+    x = Const("x", type_sort)
+    for i in range(len(funcs)):
+        # function funcs[i] will have i arguments and 1 return type, so  it uses i + 1 constants
+        consts = quantifiers_consts[:i + 1]
+        inst = funcs[i](consts)
+        axioms.append(
+            ForAll(consts, extends(inst, type_sort.func), patterns=[inst])
+        )
+        axioms.append(ForAll([x] + consts, Implies(subtype(x, inst), x == inst)))
+    return axioms
+
+
+def invert_dict(d):
+    result = {}
+    for key in d:
+        result[d[key]] = key
+
+    return result
+
+
+def classes_subtype_axioms(classes, classes_to_attrs, sub_to_base):
+    """Add the axioms for the classes subtyping"""
+    axioms = []
+    for cls in classes:
+        base_name = sub_to_base[cls]
+        if base_name == "object":
+            axioms.append(extends(classes[cls], type_sort.object))
+            continue
+
+        base = classes[base_name]
+        axioms.append(extends(classes[cls], base))
+
+    return axioms
+
+type_sort = None
+Object = zNone = Num = Complex = Float = Int = Bool = Seq = String = Bytes = None
+List = list_type = None
+Set = set_type = None
+Dict = dict_key_type = dict_value_type = None
+Tuple = Tuples = None
+Func = Funcs = None
+Classes = None
+subtype = extends = stronger_num = None
+subtype_properties = generics_axioms = num_strength_properties = axioms = None
+solver = None
+x = y = z = None
+Type = None
+All_types = {}
+Attributes = {}
+
+
+def init_types(config):
+    """Initialize the type system in Z3 using configurations given by the pre-analyzer"""
+    max_tuple_length = config["max_tuple_length"]
```

Comment:
If, for example, we have 10 or 20 configurations later, do you think it will be cleaner having an object with 10-20 attributes, and having getters/setters for them or passing their values in the object constructor?